### PR TITLE
Deprecate grammar entry "intropattern" in tactic notations in favor of "simple_intropattern"

### PR DESCRIPTION
**Pull request by herbelin**
This is to prevent confusing `intropattern` in `TACTIC EXTEND` and `Tactic Notation` with the terminology used in the grammar of tactics in the reference manual (in the former it is only what is called `simple_intropattern` in the reference manual.)

**Kind:** fixing naming

This is a proposal in reaction to the discussion at [#9288](https://github.com/coq/coq/pull/9288#issuecomment-495723706).

Note: An alternative would be to silently change the semantics of `intropattern` in `TACTIC EXTEND` and `Tactic Notation` so that it means the same as `intropattern` in the reference manual. This would probably not be a problem in practice.

### CODE REVIEW

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: More general comment about this table, it would be good if it used the token names on all rows (with the appropriate link). So this would become:

```suggestion
        - :token:`simple_intropattern`
```

and a line like:

```rst
 - qualified identifier
```

would become:

```rst
 - qualified identifier (:token:`qualid`)
```

**Review by Zimmi48 — APPROVED** [review]
I'm in favor, and the strategy you propose seem the safest, although to go faster, we could request this to be included in 8.10.0. I think this is the kind of stuff that @vbgl still accepts.

**Comment by Zimmi48** [comment]
There is an anomaly to fix though (that is raised a lot on the CI).

**Comment by herbelin** [comment]
> There is an anomaly to fix though (that is raised a lot on the CI).

This is now fixed (the printer was not declared). The other missing declaration is also fixed.

The remaining failure is with output checking in rust (would require an overlay).

I'm still wondering whether a full deprecation phase is not overkill. After all, the name `simple_intropattern` was still there for those who knew the difference. For instance, Iris uses `simple_intropattern` in their notations.

If @vbgl accepts it for 8.10, a reasonable compromise could be to have deprecation only in 8.10 and start as soon as 8.11 to have the semantics `intropattern` = `intropattern`. In particular, we can change `injection` as soon of now in 8.11.

Otherwise, I'm tempted to skip deprecation. After all, 1) the difference is only about `*` and `**` 2) it is hard to say when users used `intropattern` meaning they expected `*` and `**` to be there and when they use it thinking consciously that this would exclude `*` and `**`.

Extra note: I don't know well the details of Ltac2, but it may also make sense to use the infrastructure of Ltac2 to parse grammar entries. For instance, Ltac2 seems more comprehensive, allowing also `equality_intropattern`, `naming_intropattern`, etc.

**Comment by jfehrle** [comment]
Would this require updates to the Tactics/Intro patterns section?

I would hope all intro patterns could be documented in one place rather
than having a multiple descriptions in different chapters (e.g. both in
tactics and Ltac2).

**Comment by Zimmi48** [comment]
> If @vbgl accepts it for 8.10, a reasonable compromise could be to have deprecation only in 8.10 and start as soon as 8.11 to have the semantics `intropattern` = `intropattern`. In particular, we can change `injection` as soon of now in 8.11.

Yes, this is what I'm suggesting.

>Would this require updates to the Tactics/Intro patterns section?

I don't think so, all the intro-patterns are already documented in there. This is only changing the mapping that is used for users of Notation (and documented in the corresponding chapter).

**Comment by herbelin** [comment]
The PR needs an Iris overlay for its printing test (since `simple_intropattern` in Ltac Notation is now printed `simple_intropattern` and not anymore `intropattern`). The patch is on iris gitlab [here](https://gitlab.mpi-sws.org/herbelin/iris-coq/commits/master+consequence-coq-pr10239-intropattern-means-simple_intropattern). I can submit it upstream to iris whenever the protocol proposed to @vbgl is clear.

**Comment by herbelin** [comment]
@RalfJung: I'm afraid I still don't know how to do an overlay for Iris. I made an Iris PR for this Coq PR [here]{https://gitlab.mpi-sws.org/iris/iris/merge_requests/260).

@vbgl: this shall have to go to v8.10. OK?

**Comment by herbelin** [comment]
@Zimmi48: As far as I understood @RalfJung [here](https://gitlab.mpi-sws.org/iris/iris/merge_requests/260#note_36733) ([commit](https://gitlab.mpi-sws.org/iris/iris/commit/fad781af8fa1e927a43ea305105e2acdfa12eb23)), this can be merged without temporary breaking CI.

Rebased for double-checking. I suspect the verdi-raft failure to be independent (??).

**Comment by Zimmi48** [comment]
Can I get quickly an approval by one of the parsing maintainers? @ppedrot @ejgallego?

**Review by ppedrot — APPROVED** [review]
I guess that's OK, even I am not too fond of the addition of a new genarg.

File: plugins/ltac/tacarg.ml
Comment (ppedrot) [inline plugins/ltac/tacarg.ml]: Forgot to mention: it's probably better to share the same runtime representation by setting the `dyn` flag. Not critical though.

File: plugins/ltac/taccoerce.ml
Comment (ppedrot) [inline plugins/ltac/taccoerce.ml]: If you share the same runtime representation you can get rid of this additional check.

**Comment by Zimmi48** [comment]
I'll leave at least one more day for @herbelin to be able to see Pierre-Marie's comments.

File: plugins/ltac/tacarg.ml
Comment (herbelin) [reply plugins/ltac/tacarg.ml]: Ok, done.

File: plugins/ltac/taccoerce.ml
Comment (herbelin) [reply plugins/ltac/taccoerce.ml]: Is it possible to share the same runtime representation but different parsing rules?

**Comment by herbelin** [comment]
> I guess that's OK, even I am not too fond of the addition of a new genarg.

Is there something you suggest I should do instead?

File: plugins/ltac/taccoerce.ml
Comment (ppedrot) [reply plugins/ltac/taccoerce.ml]: Yep, `has_type` is only looking at the runtime representation.

**Comment by Zimmi48** [comment]
@herbelin The last push does not build.

**Comment by herbelin** [comment]
> The last push does not build.

I don't know what to do. Is it worth investigating why sharing the representation breaks interpretation with anomaly `"interp function not found: wit_intro_pattern"`? (Apparently, the shared representation works for `Geninterp.val_tag` but not for `Geninterp.interp`, is that expected?). In the meantime, I reverted the change.

**Comment by ppedrot** [comment]
@herbelin Geninterp.interp is bound to the genarg, not to the runtime dynamic argument. The reason is that intepretation goes from glob expressions into runtime values. Thus, sharing the same runtime tag between two genargs doesn't prevent you from having to provide an interpretation function (and intern, and subst, and whatever) for both.

**Comment by herbelin** [comment]
Does it mean that, in the balance, you recommend using a `dyn` and extra `interp`/`intern`/`subst`/... functions?

**Comment by ppedrot** [comment]
Yes. The point about sharing the same runtime representation is to reduce the number of potential coercions that have to be defined by hand when casting variables of different types.

**Comment by Zimmi48** [comment]
@ppedrot I haven't really followed the discussion in depth. Should we merge the PR in the state it is now or not? If yes, feel free to take over and do it.

**Comment by ppedrot** [comment]
You can merge as-is if you want. It won't make the genarg mess much worse, and it is unlikely to make any observable difference assuming we deal with users not trying to actively break the tactic notation mechanism.

**Comment by herbelin** [comment]
#10239 seems to have remained a bit in the middle of the ford with the code assuming that `wit_intropattern` and `wit_simple_intropattern` share the same representation of values (thanks to the `dyn` field mentioned by @ppedrot) but in practice it did not. In particular, wrt #10778, `fresh` was not anymore looking at names occuring in simple introduction patterns.

I tried sharing the "runtime dynamic argument" (i.e. same `val_tag`). See #10781.

**Pull request merged**
